examples/modernbert/modernbert.zig

```diff
@@ -0,0 +1,234 @@
+        // input, gate = self.Wi(hidden_states).chunk(2, dim=-1)
+        const input = wi_output.slice1d(-1, .{ .end = intermediate_size });
+        const gate = wi_output.slice1d(-1, .{ .start = intermediate_size });
+
+        // Apply activation function to input and multiply by gate :
+        // self.Wo(self.drop(self.act(input) * gate))
+        const activated_input = input.gelu().mul(gate);
+
+        // Perform Wo
+        return zml.call(self.Wo, .forward, .{activated_input});
+    }
+};
+
+/// Performs multi-headed self attention on a batch of unpadded sequences.
+///
+/// TODO: If Flash Attention 2 is installed, this module uses Flash Attention to improve throughput.
+/// If Flash Attention 2 is not installed, the implementation will use SDPA,
+pub const ModernBertAttention = struct {
+    Wqkv: zml.nn.Linear,
+    Wo: zml.nn.Linear,
+    is_global_attention: bool = false,
+
+    /// sdpa_attention_forward
+    pub fn forward(
+        self: ModernBertAttention,
+        hidden_states: Tensor,
+        attention_mask: Tensor,
+        sliding_window_mask: Tensor,
+        position_ids: Tensor,
+    ) Tensor {
+        const batch_size = hidden_states.shape().dim(0);
+        const seq_length = hidden_states.shape().dim(1);
+        const num_heads = 12; // config.json: num_attention_heads
+        const head_dim = 64; // config.json hidden_size / num_attention_heads = 768 / 12
+
+        if (self.is_global_attention) {
+            log.info("Global attention", .{});
+        } else {
+            log.info("Local attention", .{});
+        }
+
+        // Project to query, key, value - { batch_size, seq_len, 3 * num_heads * head_dim }
+        var qkv: Tensor = zml.call(self.Wqkv, .forward, .{hidden_states}); // Wqkv.out.0
+
+        // Reshape to { batch_size, seq_len, 3, num_heads, head_dim }
+        qkv = qkv.reshape(.{ batch_size, seq_length, 3, num_heads, head_dim });
+        // Transpose for attention computation  { batch_size, seq_len, 3, num_heads, head_dim } -> { batch_size, num_heads, 3, seq_length, head_dim }
+        qkv = qkv.transpose(.{ 0, 3, 2, 1, 4 });
+
+        // Split into query, key, value tensors - each { batch_size, num_heads, seq_length, head_dim }
+        var q = qkv.slice1d(2, .{ .start = 0, .end = 1 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
+        var k = qkv.slice1d(2, .{ .start = 1, .end = 2 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
+        var v = qkv.slice1d(2, .{ .start = 2, .end = 3 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
+
+        // Apply rotary position embeddings (RoPE)
+        // Layer 0, 3, 6, 9, 12 ... use global RoPE
+        // Layer 1, 2, 4, 5, 7, 8, 10, 11 ... use local RoPE
+        const rope_opts = zml.nn.RopeOpts{
+            .impl = .sequential,
+            .freq_base = if (self.is_global_attention) 160_000 else 10_000,
+        };
+        const pos_idx = null; // TODO: This is temporary, need to implement position_ids ?
+        _ = position_ids;
```

Comment:
you don't need to compute position id. Position id allow to compute a slice of the Rope matrix. But in a text encoder you can generate the full matrix at once, which correspond to `pos_idx = null`

I need to update the rope documenation to the new api